go/libraries/doltcore/sqle/rowexec/lookup_join.go

```diff
@@ -0,0 +1,327 @@
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/index"
+	"github.com/dolthub/dolt/go/store/pool"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+func rowIterTableLookupJoin(
+	srcIter prolly.MapIter,
+	dstIter index.SecondaryLookupIter,
+	mapping *lookupMapping,
+	rowJoiner *prollyToSqlJoiner,
+	srcFilter, dstFilter, joinFilter sql.Expression,
+	isLeftJoin bool,
+	excludeNulls bool,
+) (sql.RowIter, error) {
+	return newLookupKvIter(srcIter, dstIter, mapping, rowJoiner, srcFilter, dstFilter, joinFilter, isLeftJoin, excludeNulls)
+}
+
+type lookupJoinKvIter struct {
+	srcIter prolly.MapIter
+	srcKey  val.Tuple
+	srcVal  val.Tuple
+
+	dstIter index.SecondaryLookupIter
+	dstKey  val.Tuple
+
+	// keyTupleMapper inputs (srcKey, srcVal) to create a dstKey
+	keyTupleMapper *lookupMapping
+
+	// projections
+	joiner *prollyToSqlJoiner
+
+	// TODO convert sql.Expression to static prolly expressions
+	srcFilter  sql.Expression
+	dstFilter  sql.Expression
+	joinFilter sql.Expression
+
+	// LEFT_JOIN impl details
+	excludeNulls bool
+	isLeftJoin   bool
+	returnedARow bool
+}
+
+func (l *lookupJoinKvIter) Close(_ *sql.Context) error {
+	return nil
+}
+
+var _ sql.RowIter = (*lookupJoinKvIter)(nil)
+
+func newLookupKvIter(
+	srcIter prolly.MapIter,
+	targetIter index.SecondaryLookupIter,
+	mapping *lookupMapping,
+	joiner *prollyToSqlJoiner,
+	srcFilter, dstFilter, joinFilter sql.Expression,
+	isLeftJoin bool,
+	excludeNulls bool,
+) (*lookupJoinKvIter, error) {
+	if lit, ok := joinFilter.(*expression.Literal); ok {
+		if lit.Value() == true {
+			joinFilter = nil
+		}
+	}
+
+	return &lookupJoinKvIter{
+		srcIter:        srcIter,
+		dstIter:        targetIter,
+		joiner:         joiner,
+		keyTupleMapper: mapping,
+		srcFilter:      srcFilter,
+		dstFilter:      dstFilter,
+		joinFilter:     joinFilter,
+		isLeftJoin:     isLeftJoin,
+		excludeNulls:   excludeNulls,
+	}, nil
+}
+
+func (l *lookupJoinKvIter) Next(ctx *sql.Context) (sql.Row, error) {
+	for {
+		var err error
+		if l.dstKey == nil {
+			l.returnedARow = false
+			// get row from |src|
+			l.srcKey, l.srcVal, err = l.srcIter.Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if l.srcKey == nil {
+				return nil, io.EOF
+			}
+
+			l.dstKey, err = l.keyTupleMapper.dstKeyTuple(ctx, l.srcKey, l.srcVal)
+			if err != nil {
+				return nil, err
+			}
+
+			if err := l.dstIter.New(ctx, l.dstKey); err != nil {
+				return nil, err
+			}
+		}
+
+		dstKey, dstVal, ok, err := l.dstIter.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if !ok {
+			l.dstKey = nil
+			if !(l.isLeftJoin && !l.returnedARow) {
+				continue
+			}
+		}
+
+		ret, err := l.joiner.buildRow(ctx, l.srcKey, l.srcVal, dstKey, dstVal)
+		if err != nil {
+			return nil, err
+		}
+
+		// side-specific filters are currently hoisted
```

Comment:
Can you elaborate on this comment. It's unclear exactly what this means.

Reply:
related to the TODO with my big comment above, prolly side `sql.Expression.Eval`